### ISSUE TITLE
Unsound array copy

### ISSUE THREAD

**Issue by rodykersten** [jayhorn/jayhorn#142]
See UnsatIssue142.java

Seems like it might be related to the random length of the array.

**Comment by rodykersten** [comment]
Narrowed it down to the way that subclasses are handled in method calls. This example contains a method sort(Object[] a) which is called with a parameter of type Integer[]. If I change the signature of the method to sort(Integer[] a) the result is UNSAFE as expected.

**Comment by martinschaef** [comment]
Uh ... that might be a nasty one. Because arrays are replaced by these fake classes, it might be that we don't keep the subtype relation.... i.e., check in the Jayhorn types if this subtypes exist

**Comment by rodykersten** [comment]
That makes sense, I'll look into it, thanks

**Comment by rodykersten** [comment]
Fixed.

This was more tricky than I anticipated... Making e.g. the replacement class for Integer[] a subclass of Number[] does not work, because you'll be overriding methods and masking fields. I ended up doing this: all reference array replacements are now subclasses of the replacement for Object[]. The only method they override is the constructor, in order to set the element type; get and set + fields are inherited.
